caikit_tgis_backend/load_balancing_proxy.py

```diff
@@ -0,0 +1,169 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provides a grpc client which:
+- Sets client-side load-balancing options
+- Polls DNS and triggers channel re-connection when new endpoints are detected
+"""
+# Standard
+from functools import partial
+from typing import Generic, List, Optional, Set, Tuple, Type, TypeVar
+import socket
+import threading
+
+# Third Party
+import grpc
+
+# First Party
+from caikit.core.exceptions import error_handler
+import alog
+
+T = TypeVar("T")
+
+log = alog.use_channel("TGCONN")
+error = error_handler.get(log)
+
+
+# pylint: disable=too-many-instance-attributes
+class GRPCLoadBalancerProxy(Generic[T]):
+    """Proxies a grpc client class T, reconnecting the client when new IPs are available"""
+
+    def __init__(
+        self,
+        client_class: Type[T],
+        target: str,
+        policy: str = "round_robin",
+        poll_interval_s: float = 10,
+        credentials: Optional[str] = None,
+        channel_options: Optional[List[Tuple[str, str]]] = None,
+    ):
+        # Ensure that self._client always exists. It is required by the __getattr__ proxying
+        self._client = None
+        self.client_class = client_class
+        self.target = target
+
+        error.value_check(
+            "<TGB54435438E>",
+            target.count(":") == 1,
+            "Target must be provided in {host}:{port} format",
+        )
+        channel_options = channel_options or []
+        # pylint: disable=line-too-long
+        # Cite: https://grpc.github.io/grpc/core/group__grpc__arg__keys.html#ga72c2b475e218ecfd36bb7d3551d0295b
+        channel_options.append(("grpc.lb_policy_name", policy))
+
+        # Save a partial for re-constructing channels later
+        if credentials:
+            log.debug3("Creating load-balancing client with secure channel")
+            self.channel_partial = partial(
+                grpc.secure_channel,
+                target=self.target,
+                options=channel_options,
+                credentials=credentials,
+            )
+        else:
+            log.debug3("Creating load-balancing client with insecure channel")
+            self.channel_partial = partial(
+                grpc.insecure_channel, target=self.target, options=channel_options
+            )
+
+        # Build the client once
+        self._client = self.client_class(self.channel_partial())
+
+        # Get initial IP set
+        self._ip_set: Set[Tuple[str, int]] = set()
+
+        self.poll_interval = poll_interval_s
+        self._timer: Optional[threading.Timer] = None
+        self._poll_lock = threading.Lock()
+        self._shutdown = False
+        self._dns_poll()
```

Comment:
I think this would be a useful feature but potentially for different cluster setup. This sounds like a feature request we can address later? (I can create an issue if that sounds good)

Reply:
I actually want to get this in now and I think it's as simple as changing the above `value_check` to `>=0` and then putting an `if` here. The real reason I want to get this in is that I'm concerned there may be corners we're missing in review and I want to be able to disable the polling portion if they come up.